### PR TITLE
[feat]: [OIP-588]: RegEx test node and control node pattern

### PR DESCRIPTION
**Pull request by AbhijithHarness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50207)

### CODE REVIEW

**Comment by AbhijithHarness** [comment]
trigger uts

**Comment by AbhijithHarness** [comment]
trigger commonchecks

**Comment by AbhijithHarness** [comment]
trigger uts

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/verificationjob/utils/VerificationJobInstanceDataCollectionUtils.java [hunk 1]
```diff
@@ -27,14 +29,37 @@ public boolean shouldCollectPreDeploymentData(VerificationJobInstance verificati
...
+    }
+    return Arrays.asList(verificationJobInstance.getServiceInstanceDetails().getControlNodeRegExPattern());
+  }
+
+  public List<String> validPostDeploymentNodePatterns(VerificationJobInstance verificationJobInstance) {
+    if (StringUtils.isEmpty(verificationJobInstance.getServiceInstanceDetails().getTestNodeRegExPattern())) {
+      return null;
+    }
+    if (verificationJobInstance.getResolvedJob().getType().equals(VerificationJobType.ROLLING)
+        || verificationJobInstance.getResolvedJob().getType().equals(VerificationJobType.BLUE_GREEN)) {
```
Comment (dhruv-harness) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/verificationjob/utils/VerificationJobInstanceDataCollectionUtils.java:50]: should we not check for load test, simple verification types here? maybe we can put the check on canary and auto, and keep default behaviour for others

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/cdng/services/impl/CVNGStep.java [hunk 2]
```diff
@@ -70,12 +67,7 @@
 import java.time.Clock;
 import java.time.Duration;
 import java.time.Instant;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.List;
-import java.util.Map;
-import java.util.Objects;
-import java.util.UUID;
+import java.util.*;
```
Comment (dhruv-harness) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/cdng/services/impl/CVNGStep.java:70]: nit: wrong import type

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/cdng/beans/HostLevelVerificationJobSpec.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import lombok.NoArgsConstructor;
+import lombok.experimental.SuperBuilder;
+
+@SuperBuilder
+@NoArgsConstructor
+public abstract class HostLevelVerificationJobSpec extends VerificationJobSpec {
+  @ApiModelProperty(dataType = SwaggerConstants.BOOLEAN_CLASSPATH, value = "Possible values: [true, false]")
+  ParameterField<Boolean> shouldUseCDNodes;
+
+  @ApiModelProperty(dataType = SwaggerConstants.STRING_CLASSPATH) ParameterField<String> testNodeRegExPattern;
```
Comment (n00bitax) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/cdng/beans/HostLevelVerificationJobSpec.java:23]: can we validate if the pattern is correct, the moment we enter the verify step

**Comment by AbhijithHarness** [comment]
trigger uts

**Comment by AbhijithHarness** [comment]
trigger commonchecks

**Comment by AbhijithHarness** [comment]
trigger uts

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/verificationjob/utils/VerificationJobInstanceDataCollectionUtils.java (hunk 1 above)
Comment (AbhijithHarness) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/verificationjob/utils/VerificationJobInstanceDataCollectionUtils.java:52]: here the default behaviour should be same as AUTO and CANARY, where both control and test node regex are added

Comment (AbhijithHarness) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/verificationjob/utils/VerificationJobInstanceDataCollectionUtils.java:52]: For Load test, we dont collect node level information anyways, right?

**Comment by AbhijithHarness** [comment]
trigger uts

**Comment by AbhijithHarness** [comment]
trigger uts

**Comment by n00bitax** [comment]
LGTM

**Review by n00bitax — APPROVED** [review]

**Pull request merged**